File: benchmarks/solutions/ruby/a3-valid-parens.rb
def valid?(s)
  stack = []
  pairs = { ")" => "(", "]" => "[", "}" => "{" }
  s.each_char do |c|
    if pairs.key?(c)
      return false unless stack.pop == pairs[c]
    else
      stack.push(c)
    end
  end
  stack.empty?
end

File: benchmarks/solutions/javascript/a3-valid-parens.js
function isValid(s) {
  const stack = [];
  const pairs = { ")": "(", "]": "[", "}": "{" };
  for (const c of s) {
    if (c in pairs) {
      if (stack.pop() !== pairs[c]) return false;
    } else {
      stack.push(c);
    }
  }
  return stack.length === 0;
}
